language/tools/move-bytecode-viewer/src/bytecode_viewer.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+use crate::interfaces::LeftScreen;
+use bytecode_source_map::{mapping::SourceMapping, source_map::SourceMap};
+use disassembler::disassembler::{Disassembler, DisassemblerOptions};
+use move_ir_types::location::Loc;
+use regex::Regex;
+use std::collections::HashMap;
+use vm::{
+    access::ModuleAccess,
+    file_format::{CodeOffset, CompiledModule, FunctionDefinitionIndex},
+};
+
+#[derive(Clone, Debug)]
+pub struct BytecodeInfo {
+    pub function_name: String,
+    pub function_index: FunctionDefinitionIndex,
+    pub code_offset: CodeOffset,
+}
+
+#[derive(Clone, Debug)]
+pub struct BytecodeViewer {
+    pub lines: Vec<String>,
+    pub module: CompiledModule,
+    pub line_map: HashMap<usize, BytecodeInfo>,
+}
+
+impl BytecodeViewer {
+    pub fn new(source_map: SourceMap<Loc>, module: CompiledModule) -> Self {
+        let source_mapping = SourceMapping::new(source_map, module.clone());
+        let mut options = DisassemblerOptions::default();
+        options.print_code = true;
+        options.print_basic_blocks = true;
+        let disassembled_string = Disassembler::new(source_mapping, options)
+            .disassemble()
+            .unwrap();
+
+        let mut base_viewer = Self {
+            lines: disassembled_string.lines().map(|x| x.to_string()).collect(),
+            line_map: HashMap::new(),
+            module,
+        };
+        base_viewer.build_mapping();
+        base_viewer
+    }
+
+    fn build_mapping(&mut self) {
+        let regex = Regex::new(r"^(\d+):.*").unwrap();
+        let fun_regex = Regex::new(r"^public\s+([a-zA-Z_]+)\(").unwrap();
+        let mut current_fun = None;
+        let mut current_fdef_idx = None;
+        let mut line_map = HashMap::new();
+
+        for (i, line) in self.lines.iter().enumerate() {
+            let line = line.trim();
+            if let Some(cap) = fun_regex.captures(line) {
+                let fn_name = cap.get(1).unwrap().as_str();
+                let function_definition_index = self
+                    .module
+                    .function_defs()
+                    .iter()
+                    .position(|fdef| {
+                        let name = self
+                            .module
+                            .identifier_at(self.module.function_handle_at(fdef.function).name);
+                        name.as_str() == fn_name
+                    })
```

Comment:
I would just eagerly build Identifier => fdef_idx mapping, to prevent this O(n^2) behavior